zkir-assembler/src/parser.rs

```diff
@@ -68,11 +68,404 @@ pub fn parse_register(name: &str) -> Result<Register> {
+            Ok(Instruction::Remu { rd, rs1, rs2 })
+        }
+
+        // R-type: Logic
+        "and" => {
+            let (rd, rs1, rs2) = parse_r_type(mnemonic, operands)?;
+            Ok(Instruction::And { rd, rs1, rs2 })
+        }
+        "or" => {
+            let (rd, rs1, rs2) = parse_r_type(mnemonic, operands)?;
+            Ok(Instruction::Or { rd, rs1, rs2 })
+        }
+        "xor" => {
+            let (rd, rs1, rs2) = parse_r_type(mnemonic, operands)?;
+            Ok(Instruction::Xor { rd, rs1, rs2 })
+        }
+        "sll" => {
+            let (rd, rs1, rs2) = parse_r_type(mnemonic, operands)?;
+            Ok(Instruction::Sll { rd, rs1, rs2 })
+        }
+        "srl" => {
+            let (rd, rs1, rs2) = parse_r_type(mnemonic, operands)?;
+            Ok(Instruction::Srl { rd, rs1, rs2 })
+        }
+        "sra" => {
+            let (rd, rs1, rs2) = parse_r_type(mnemonic, operands)?;
+            Ok(Instruction::Sra { rd, rs1, rs2 })
+        }
+        "slt" => {
+            let (rd, rs1, rs2) = parse_r_type(mnemonic, operands)?;
+            Ok(Instruction::Slt { rd, rs1, rs2 })
+        }
+        "sltu" => {
+            let (rd, rs1, rs2) = parse_r_type(mnemonic, operands)?;
+            Ok(Instruction::Sltu { rd, rs1, rs2 })
+        }
+
+        // I-type: Arithmetic Immediate
+        "addi" => {
+            let (rd, rs1, imm) = parse_i_type(mnemonic, operands)?;
+            Ok(Instruction::Addi { rd, rs1, imm })
+        }
+        "slti" => {
+            let (rd, rs1, imm) = parse_i_type(mnemonic, operands)?;
+            Ok(Instruction::Slti { rd, rs1, imm })
+        }
+        "sltiu" => {
+            let (rd, rs1, imm) = parse_i_type(mnemonic, operands)?;
+            Ok(Instruction::Sltiu { rd, rs1, imm })
+        }
+        "xori" => {
+            let (rd, rs1, imm) = parse_i_type(mnemonic, operands)?;
+            Ok(Instruction::Xori { rd, rs1, imm })
+        }
+        "ori" => {
+            let (rd, rs1, imm) = parse_i_type(mnemonic, operands)?;
+            Ok(Instruction::Ori { rd, rs1, imm })
+        }
+        "andi" => {
+            let (rd, rs1, imm) = parse_i_type(mnemonic, operands)?;
+            Ok(Instruction::Andi { rd, rs1, imm })
+        }
+
+        // I-type: Shift Immediate (special handling for shamt)
+        "slli" => {
+            let (rd, rs1, imm) = parse_i_type(mnemonic, operands)?;
+            let shamt = parse_shamt(&imm.to_string())?;
+            Ok(Instruction::Slli { rd, rs1, shamt })
+        }
+        "srli" => {
+            let (rd, rs1, imm) = parse_i_type(mnemonic, operands)?;
+            let shamt = parse_shamt(&imm.to_string())?;
+            Ok(Instruction::Srli { rd, rs1, shamt })
+        }
+        "srai" => {
+            let (rd, rs1, imm) = parse_i_type(mnemonic, operands)?;
+            let shamt = parse_shamt(&imm.to_string())?;
+            Ok(Instruction::Srai { rd, rs1, shamt })
+        }
```

Comment:
Shift instruction parsing is inefficient. Lines 371-385 parse the immediate as i16 through 'parse_i_type()', then convert it back to string just to validate it as a shift amount. The third operand should be parsed directly as shamt instead of going through the i16 conversion first.